### PR TITLE
SPIN-1478 & SPIN- 1484: This refactoring of the account - region(zones) - cluster selector.

### PR DESCRIPTION
**Pull request by zanthrash**
This changes is for showing ALL the regions (zones for gce) when the free-form text input is selected for the cluster picker for pipeline stages.

This is to allow a user to set a region and cluster for an app that dosn't yet have the cluster deployed.

It also reduce the amount of widget specific logic that is repeated in all the stage controllers and moves it to one place.

@anotherchrisberry Please review and maybe pull down and kick the tires
@duftler I made an attempt to add this to all the GCE stages with a Zone specific directive.  Please take a look and let me know if I broke your stage configs.

### CODE REVIEW

File: app/scripts/modules/core/pipeline/config/stages/disableAsg/aws/disableAsgStage.html [hunk 1]
```diff
@@ -1,25 +1,32 @@
...
-        on-change="disableAsgStageCtrl.resetSelectedCluster()">
-      </checklist>
-    </stage-config-field>
-    <stage-config-field label="Cluster" help-key="pipeline.config.disableAsg.cluster">
-      <cluster-selector
-        clusters="clusterList"
-        model="stage.cluster">
-      </cluster-selector>
-    </stage-config-field>
+<h2>boom</h2>
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/pipeline/config/stages/disableAsg/aws/disableAsgStage.html]: boom

File: app/scripts/modules/core/widgets/accountRegionClusterSelector.component.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
+'use strict';
+
+let angular = require('angular');
+
+module.exports = angular
+  .module('spinnaker.core.accountRegionClusterSelector.directive', [
+    require('../../core/application/listExtractor/listExtractor.service'),
+  ])
+  .directive('accountRegionClusterSelector', function() {
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/widgets/accountRegionClusterSelector.component.js]: I see the file is named `accountRegionClusterSelector.component.js` - did you want to write this as a `component`?

File: app/scripts/modules/core/widgets/accountZoneClusterSelector.component.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        };
+
+        let setClusterList = () => {
+          let clusterFilter = appListExtractorService.clusterFilterForCredentialsAndZone(vm.component.credentials, vm.component.zones);
+          vm.clusterList = appListExtractorService.getClusters([vm.application], clusterFilter);
+        };
+
+        vm.accountChange = () => {
+          vm.accountUpdated();
+        };
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/widgets/accountZoneClusterSelector.component.js]: This method seems unnecessary, since it just calls `accountUpdated`

**Comment by anotherchrisberry** [comment]
A few little things but overall looks good - cleans up a lot of boilerplate code

File: app/scripts/modules/core/widgets/accountRegionClusterSelector.component.js (hunk 2 above)
Comment (zanthrash) [inline app/scripts/modules/core/widgets/accountRegionClusterSelector.component.js]: Yeah.  I started off that way.  Will do it once I learn it. Fixing the others

**Comment by anotherchrisberry** [comment]
LGTM - needs a rebase

**Comment by duftler** [comment]
Sorry, late getting to this, but having a some troubles getting it to work ok for GCE. About to capture some screen-grabs.

**Comment by zanthrash** [comment]
@duftler let me know, we'll get it working for you.

**Comment by duftler** [comment]
Thanks @zanthrash. Have worked through this, and here's what I see:

When I configure an Amazon Find Image, it first comes up with the open box for entering a cluster name. All regions are listed (even ones not configured for the account), not just regions within which I have server groups deployed.
image

If I click "Toggle for list of existing clusters", I get the pulldown cluster list with just the names of the clusters that have server groups deployed to the selected regions. Just the regions which have deployed server groups are listed.
image

When I configure a Google Find Image, it first comes up with the open box for entering a cluster name. All zones are listed, not just zones within which I have server groups deployed.
image

If I click "Toggle for list of existing clusters", I get the pulldown cluster list with just the names of the clusters that have server groups deployed to the selected zones. But all the zones are still listed.
image

If I click "Toggle for text input", I now get just the zones in one region (us-central1), no matter where I have server groups deployed.
image

**Comment by duftler** [comment]
Still experimenting, but I see the same behavior with each of the stages that has the cluster selector.

**Comment by zanthrash** [comment]
@duftler Yeah it'll be the same for each stage.  From your screengrabs it looks like the AWS functionally is working correctly.  Can you elaborate on the behavior you are expecting?

I'm taking a second look at the logic in the accountZoneClusterSelector.component.js.

**Pull request merged**